tools/harness-automation/parse_topofile.py

```diff
@@ -0,0 +1,123 @@
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+
+"""
+parse_topofile.py
+-----------------
+This script is used to parse TopologyConfig.txt file to list vendor device info
+when preparing for the Thread Certification testbed
+"""
+
+import argparse
+import logging
+import re
+
+logging.basicConfig(format='%(message)s', level=logging.DEBUG)
+MAX_VENDOR_DEVICE = 32
+
+
+def device_calculate(topo_file, case_list):
+    try:
+        case_count = 0
+        testbed_vendor_dict = {}
+        for case in case_list:
+            f = open(topo_file, 'r')
+            while 1:
+
+                case_vendor_dict = {}
+
+                line = f.readline()
+                line = line.strip()
+                if not line:
+                    break
+                if re.match(r'#.*', line):
+                    continue
+
+                matched_case = re.match(r'(.*)-(.*)', line, re.M | re.I)
+
+                if case != 'all' and case != matched_case.group(1):
+                    continue
+
+                logging.info('case %s:' % matched_case.group(1))
+                case_count += 1
+                role_vendor_str = matched_case.group(2)
+                role_vendor_raw_list = re.split(',', role_vendor_str)
+                role_vendor_list = []
+
+                for i in range(len(role_vendor_raw_list)):
+                    device_pair = re.split(':', role_vendor_raw_list[i])
+                    role_vendor_list.append(tuple(device_pair))
+                logging.info('\trole-vendor pair: %s' % role_vendor_list)
+
+                for i in range(len(role_vendor_list)):
+                    if role_vendor_list[i][1] in case_vendor_dict:
+                        case_vendor_dict[role_vendor_list[i][1]] += 1
+                    else:
+                        case_vendor_dict[role_vendor_list[i][1]] = 1
+
+                    if role_vendor_list[i][1] in testbed_vendor_dict:
+                        if testbed_vendor_dict[role_vendor_list[i][1]] < case_vendor_dict[role_vendor_list[i][1]]:
+                            testbed_vendor_dict[role_vendor_list[i][1]] = case_vendor_dict[role_vendor_list[i][1]]
+                    else:
+                        testbed_vendor_dict[role_vendor_list[i][1]] = 1
+
+                logging.info('\tneeded vendor devices:%s' % case_vendor_dict)
+            f.close
+
+        count_any = MAX_VENDOR_DEVICE
+        for key in testbed_vendor_dict:
+            if key != 'Any':
+                count_any -= testbed_vendor_dict[key]
+        if 'Any' in testbed_vendor_dict:
+            testbed_vendor_dict['Any'] = count_any
```

Comment:
This calculation of `Any` assumes that `Any` is only used by 5.2.3 which has 32 devices?

Reply:
yes, 'Any' = 10 in case 5.2.3. When setting up the testbed for all cases, every vendor's device number is the maximum needed device number in each case. Thus, 'Any' will be 32 - (all vendor devices number)